cookbooks/ruby/libraries/helpers.rb

```diff
@@ -0,0 +1,176 @@
+module RubyHelpers
+  def ruby2x?(x, version_string)
+    version_string =~ /^2\.#{x}\.[\d\.]*\b/
+  end
+
+  def ruby_minor_version(version_string)
+    version_string.split('.')[1]
+  end
+
+  def ruby2x_mask(version_string)
+    "-ruby_targets_ruby2#{ruby_minor_version(version_string)}"
+  end
+
+  def get_installed_ruby2x_slots
+    cmd = Mixlib::ShellOut.new("equery l -F '$slot' dev-lang/ruby")
+    cmd.run_command.stdout.split
+  end
+
+  def get_nondefault_ruby2x_slots
+    get_installed_ruby2x_slots.select do |slot|
+      minor = slot.split('.')[1].to_i
+      minor > 3 # everything above ruby 2.3 is not default
+    end
+  end
+
+  def get_nondefault_ruby2x_use_flags
+    slots = get_nondefault_ruby2x_slots
+    slots.map { |slot| "ruby_targets_ruby#{slot.sub('.', '')}" }
+  end
+
+  def need_ruby2x_reset
+    # This is how we decide if the Ruby packages need to be reset:
+    # 1. Get the installed Ruby slots that are above 2.3 (e.g. 2.4, 2.5, ...)
+    # 2. if the highest installed slot is above 2.3 and the desired version is below that,
+    # we need to reset the Ruby packages
+    slots = get_nondefault_ruby2x_slots
+    return false if slots.empty? # highest installed ruby version is not a nondefault version -> everything's fine
+    max_installed_minor = slots.map { |slot| slot.split('.')[1].to_i }.max
+    component = node.engineyard.environment.ruby
+    desired_minor = ruby_minor_version(component[:version]).to_i
+    if desired_minor < max_installed_minor
+      return true # we are installing a lower version than the already installed
+    else
+      return false # we don't change the version, or upgrade -> fine
+    end
+  end
+
+  def get_nondefault_installed_ruby2x_deps
+    flags = get_nondefault_ruby2x_use_flags
+    flags.flat_map do |flag|
+      cmd = Mixlib::ShellOut.new("equery h #{flag}")
+      cmd.run_command.stdout.split.map { |pkg| "=#{pkg}" }
+    end
+  end
+
+  def get_nondefault_installed_ruby2x
+    cmd = Mixlib::ShellOut.new("equery l -F '=$category/$name-$version' \\>=dev-lang/ruby-2.4") # everything above Ruby 2.3
+    cmd.run_command.stdout.split
+  end
+
+  def get_nondefault_ruby_packages
+    get_nondefault_installed_ruby2x_deps + get_nondefault_installed_ruby2x
+  end
+
+  def remove_nondefault_ruby_packages
+    packages = get_nondefault_ruby_packages
+    if not packages.empty?
+      execute 'remove non-default ruby packages' do
+        command %Q{emerge -Cv --color n --nospinner --quiet #{packages.join(' ')}}
+        action :run
+      end
+    end
+  end
+
+  def reset_ruby_mask
+    component = node.engineyard.environment.ruby
+    ruby_version = component[:version]
+    ruby_mask = ruby2x_mask(ruby_version)
+    use_mask_clear "ruby" do
+      mask_file "ruby"
+    end
+    use_mask ruby_mask do
+      mask_file "ruby"
+      only_if "ruby_mask"
```

Comment:
I think this needs to be `only_if { ruby_mask }`. Although, ruby2x_mask(ruby_version) always return a string so it's always true.

Reply:
Agree, I will change it.

Reply:
That's also the case in `cookbooks/ruby/recipes/ruby.rb`. Will change it there too.